src/lib.rs

```diff
@@ -3921,6 +3925,61 @@ pub struct SemanticTokensLegend {
     pub token_modifiers: Vec<SemanticTokenModifier>,
 }
 
+/**
+ * The actual tokens. For a detailed description about how the data is
+ * structured please see
+ * https://github.com/microsoft/vscode-extension-samples/blob/5ae1f7787122812dcc84e37427ca90af5ee09f14/semantic-tokens-sample/vscode.proposed.d.ts#L71
+ */
+#[derive(Debug, Eq, PartialEq, Copy, Clone, Default)]
+#[cfg(feature = "proposed")]
+pub struct SemanticToken {
+    pub delta_line: u32,
+    pub delta_start: u32,
+    pub length: u32,
+    pub token_type: u32,
+    pub token_modifiers_bitset: u32,
+}
+
+#[cfg(feature = "proposed")]
+impl SemanticToken {
+    fn deserialize_tokens<'de, D>(deserializer: D) -> Result<Vec<SemanticToken>, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let data = Vec::<u32>::deserialize(deserializer)?;
```

Comment:
Nit: It is actually not that much boilerplate to write out the `Visitor` directly. Would avoid this intermediate `Vec`.

https://github.com/serde-rs/serde/blob/142955b109ed70fe67d9a7dc28bb7403ad0dd041/serde/src/de/impls.rs#L846-L877

Reply:
Can you elaborate on this? I'm not quite sure what you're proposing. In my head I want an implementation of Serialize/Deserialize for `SemanticToken` that when used in a `Vec<SemanticToken>` for ser and `Vec<u32>` for de packs everything up properly but I find serde to be tricky.

Reply:
I meant that you can do something like this to directly create the `Vec<SemanticToken>` and avoid the intermediate `Vec<u32>`

```rust
        struct VecVisitor;

        impl<'de> Visitor<'de> for VecVisitor
        {
            type Value = Vec<SemanticToken>;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("a sequence")
            }

            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
            where
                A: SeqAccess<'de>,
            {
                let mut values = Vec::with_capacity(size_hint::cautious(seq.size_hint() /  5));

                while let Some(a) = try!(seq.next_element()) {
                    let a = try!(seq.next_element()) {
                        Some(a) => a,
                        None => // Not enough elements
                    }
                    // Extract 3 more elements
                    values.push(SemanticToken { ... });
                }

                Ok(values)
            }
        }

        let visitor = VecVisitor;
        deserializer.deserialize_seq(visitor)
```

Reply:
@Marwes I think someone more familiar with serde should write the visitor. I tried what you recommended but it ended up getting very ugly.